service_change.go

```diff
@@ -30,15 +37,72 @@ func (w *ServiceChangeWaiter) RefreshFunc() resource.StateRefreshFunc {
 			return nil, "", err
 		}
 
-		return service, service.State, nil
+		// If this is a Kafka service, wait for its component parts (i.e. Kafka Connect,
+		// Kafka REST, and Schema Registry) to be ready
+		state := service.State
+		if !kafkaServicesReady(service) {
+			state = aivenKafkaServicesStartingState
+		}
+
+		return service, state, nil
+	}
+}
+
+// If any of Kafka Rest, Schema Registry, and Kafka Connect are enabled, refresh
+// their state to check if they're ready
+func kafkaServicesReady(service *aiven.Service) bool {
+	// Check if the service is a Kafka service and Kafka itself is ready
+	if service.Type != "kafka" {
+		return true
+	}
+	if service.State != aivenTargetState {
+		return false
+	}
+
+	kafkaRestEnabled, schemaRegistryEnabled, kafkaConnectEnabled := false, false, false
+	userConfig := service.UserConfig.(map[string]interface{})
+	if enabled, ok := userConfig["kafka_rest"]; ok {
+		kafkaRestEnabled = enabled.(bool)
 	}
+	if enabled, ok := userConfig["schema_registry"]; ok {
+		schemaRegistryEnabled = enabled.(bool)
+	}
+	if enabled, ok := userConfig["kafka_connect"]; ok {
+		kafkaConnectEnabled = enabled.(bool)
+	}
+
+	if !kafkaRestEnabled && !schemaRegistryEnabled && !kafkaConnectEnabled {
+		return true
+	}
+
+	// Ping each Kafka add-on service's url to see if they are alive
+	if kafkaRestEnabled {
+		if resp, err := http.Get(service.ConnectionInfo.KafkaRestURI); err != nil {
+			resp.Body.Close()
```

Comment:
`resp.Body.Close()` should be called outside of the if statement. Since we need to close it if `err` is `nil`.

You could also create a utility method that encapsulate all this logic and simply takes a URI as param.

I would also get rid of some of the booleans in here. It would look something like this:

```
func kafkaRestReady(service *aiven.Service) bool {
 	if enabled, ok := userConfig["kafka_rest"]; ok && enabled.(bool) {
 	 	return uriReachable(service.ConnectionInfo.KafkaRestURI)
 	}
       return true
}

func uriReachable(uri string) bool {
       resp, err := http.Get(uri)
       defer resp.Body.Close()
       return err == nil
}
```